lib/materialize.nix

```diff
@@ -81,15 +102,19 @@ let
   calculateUseHash = derivation.overrideAttrs (_: hashArgs);
   calculateUseAll = 
     # Skip right to expectedPath if it already exists
-    if materialized != null && builtins.pathExists materialized
-      then runCommand name hashArgs ''
-        cp -r ${materialized} $out
-        # Make sure output files can be removed from the sandbox
-        chmod -R +w $out
-      ''
+    if materialized != null
+      then
+        assert __pathExists materialized;
+        runCommand name hashArgs ''
+          cp -r ${materialized} $out
+          # Make sure output files can be removed from the sandbox
+          chmod -R +w $out
+        ''
       else calculateUseHash;
 
 in
-  if checkMaterialization
+  # Use the checked version if requested or if the `materialized` version
+  # is missing (perhaps deleted or not created yet).
+  if checkMaterialization || (materialized != null && !__pathExists materialized)
```

Comment:
I think this would be clearer if we distribute the not out of the second condition. Something like `let usableMaterialized = materialized != null && __pathExists materialized in checkMaterialization || !usableMaterialized`

Reply:
That is slightly different though.  We only want to force the checked path when the materialized was specified ( `!= null` ), but it does not exist on disk.  If materialized is not specified at all ( `==null` ) we don't want to force a full build to check the hash.  I have added a `missingMaterialized` and a comment to try to make the distinction clearer.